tests/typecheck_fail.rs

```diff
@@ -230,3 +230,10 @@ fn piecewise_signature() {
         Err(TypecheckError::TypeMismatch(..))
     );
 }
+#[test]
+fn recursive_let() {
+    assert_matches!(
+        type_check_expr("let rec f : Num -> Num = fun x => f \"hoi\" in null"),
+        Err(TypecheckError::TypeMismatch(..))
+    );
+}
```

Comment:
It's good to add failing tests, but passing tests are nice too :slightly_smiling_face: I would suggest to add a few tests in `pass/typechecking.ncl` to check that recursive functions are correctly typechecked, including polymorphic recursive functions, something like:

```nickel
let rec f : forall a. a -> Num -> a = fun x n => if n == 0 then a else if f "0" n == "1" then f x (n - 1) else f x (f 1 n)
```